### PR TITLE
Migrate part of rustc_infer to session diagnostic

### PR DESCRIPTION
**Pull request by IntQuant**

### CODE REVIEW

**Comment by IntQuant** [comment]
@rustbot label +A-translation

r? rust-lang/diagnostics
cc https://github.com/rust-lang/rust/issues/100717

File: compiler/rustc_infer/src/lib.rs [hunk 1]
```diff
@@ -23,6 +23,8 @@
 #![feature(never_type)]
 #![feature(try_blocks)]
 #![recursion_limit = "512"] // For rustdoc
+// #![deny(rustc::untranslatable_diagnostic)]
+// #![deny(rustc::diagnostic_outside_of_impl)]
```
Comment (davidtwco) [inline compiler/rustc_infer/src/lib.rs:27]: What is still remaining that prevents these from being removed?

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 2]
```diff
@@ -2879,6 +2880,23 @@ impl<'tcx> ObligationCauseExt<'tcx> for ObligationCause<'tcx> {
             _ => "types are compatible",
         }
     }
+
+    fn as_requirement_localised(&self) -> &'static str {
+        use crate::traits::ObligationCauseCode::*;
+        match self.code() {
+            CompareImplItemObligation { kind: ty::AssocKind::Fn, .. } => "method_compat",
```
Comment (davidtwco) [inline compiler/rustc_infer/src/infer/error_reporting/mod.rs:2887]: Would it be possible to transform these into the strings-to-be-matched-against in an `IntoDiagnosticArg` impl or something like that? Similarly for other types you do this with.

**Review by davidtwco — CHANGES_REQUESTED** [review]
This is a great start, thanks for working on it, I've left a couple comments.

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs (hunk 2 above)
Comment (IntQuant) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs:2887]: Noted, will do.

File: compiler/rustc_infer/src/lib.rs (hunk 1 above)
Comment (IntQuant) [reply compiler/rustc_infer/src/lib.rs:27]: There is still quite a bit to do - about 60 lints to resolve(54 of which are in infer/error_reporting/mods.rs). Also sometimes this api is abstracted in a function, like `note_and_explain_region`, so there is even more than that.

Comment (davidtwco) [reply compiler/rustc_infer/src/lib.rs:27]: Do you want to land some of this work before the crate is complete or just keep going until it's all done?

Comment (IntQuant) [reply compiler/rustc_infer/src/lib.rs:27]: Should be fine to merge.

Comment (davidtwco) [reply compiler/rustc_infer/src/lib.rs:27]: Could we remove these commented lines before doing so?

**Review by davidtwco — APPROVED** [review]

**Pull request merged**